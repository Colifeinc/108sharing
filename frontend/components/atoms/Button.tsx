import * as React from 'react'
import { ReactNode } from 'react'
import classNames from 'classnames'

interface IProps {
  id?: string
  handleClick?: () => void
  type?: 'submit' | 'button' | 'reset'
  size?: 'small' | 'midium'
  buttonType?: 'primary' | 'primary-filled' | 'secondary' | 'tertiary' | 'muted'
  fit?: boolean
  full?: boolean
  link?: string
  disabled?: boolean
  className?: string
  iconLeft?: ReactNode
  iconRight?: ReactNode
}

const classes = {
  size: {
    small: 'w-48 max-w-full h-10 text-sm',
    medium: 'w-64 max-w-full h-12 text-base',
  },
  buttonType: {
    primary: 'bg-gradient-primary text-white',
    'primary-filled': 'bg-primary text-white',
    secondary: 'bg-white border border-primary text-primary',
    tertiary: 'bg-white border border-blue text-blue',
    muted: 'bg-neutral-400',
  },
} as const

const Button: React.FC<IProps> = ({
  handleClick = () => {},
  disabled = false,
  full = false,
  fit = false,
  size = 'medium',
  buttonType = 'primary',
  // type = 'button',
  type = 'submit',
  className = '',
  link = null,
  iconLeft,
  iconRight,
  ...props
}) => {
  const Tag = link ? 'a' : 'button'
  return (
    <Tag
      className={classNames([
        'flex items-center justify-center rounded-full font-bold shadow-button focus:outline-none',
        classes.size[size],
        !disabled && classes.buttonType[buttonType],
        disabled && 'bg-none bg-neutral-300 text-neutral-500 cursor-auto',
        (iconLeft || iconRight) && 'relative',
        full && 'w-full max-w-none',
        fit ? 'px-4 w-auto max-w-none' : '',
        className,
      ])}
      type={link ? undefined : type}
      disabled={link ? undefined : disabled}
      onClick={handleClick}
      href={link || undefined}
      {...props}
    >
      {iconLeft && <span className="mr-2">{iconLeft}</span>}
      <span>{props.children}</span>
      {iconRight && <span className="ml-2">{iconRight}</span>}
    </Tag>
  )
}

export default Button
