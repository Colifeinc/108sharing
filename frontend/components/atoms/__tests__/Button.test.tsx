import { shallow } from 'enzyme'
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import Button from '../Button'

describe('<Button />', () => {
  it('matches to the snapshot', () => {
    const component = renderer.create(<Button />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('call handleClick props', () => {
    const mock = jest.fn()
    const wrapper = shallow(<Button handleClick={mock} />)
    wrapper.simulate('click')
    expect(mock).toHaveBeenCalled()
  })

  it('renders children when passed in', () => {
    const wrapper = shallow(
      <Button>
        <div className="unique" />
      </Button>
    )
    expect(wrapper.contains(<div className="unique" />)).toBe(true)
  })
})
