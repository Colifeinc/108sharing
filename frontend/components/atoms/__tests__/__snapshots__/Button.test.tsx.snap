// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches to the snapshot 1`] = `
<button
  className="flex items-center justify-center rounded-full font-bold shadow-button focus:outline-none w-64 max-w-full h-12 text-base bg-gradient-primary text-white"
  disabled={false}
  onClick={[Function]}
  type="submit"
>
  <span />
</button>
`;
